File: Pwr/Sem4/JezykiIParadygmatyProgramowania/Lista6/zad1/ocaml/totient1.ml
let rec gcd' a b =
  if b = 0 then a
  else gcd' b (a mod b)

let coprime a b =
  gcd' a b = 1

let rec count i n acc =
  if i > n then acc
  else
    let acc =
      if coprime i n then acc + 1 else acc
    in
    count (i + 1) n acc

let totient n =
  count 1 n 0

let () =
  let n = int_of_string Sys.argv.(1) in
  Printf.printf "%d\n" (totient n)

File: Pwr/Sem4/JezykiIParadygmatyProgramowania/Lista6/zad1/haskel/totient1.hs
import System.Environment (getArgs)

gcd' :: Integer -> Integer -> Integer
gcd' a 0 = a
gcd' a b = gcd' b (a `mod` b)

coprime :: Integer -> Integer -> Bool
coprime a b = gcd' a b == 1

totient :: Integer -> Integer
totient n = toInteger (length [k | k <- [1..n], coprime k n])

main :: IO ()
main = do
    args <- getArgs
    let n = read (head args)
    print (totient n)